modules/python/python-tf.c

```diff
@@ -99,19 +99,25 @@ _py_invoke_template_function(PythonTfState *state, const gchar *function_name, L
 
 /* NOTE: consumes ret */
 static gboolean
-_py_convert_return_value_to_result(const gchar *function_name, PyObject *ret, GString *result)
+_py_convert_return_value_to_result(PythonTfState *state, const gchar *function_name, PyObject *ret, GString *result,
+                                   LogMessageValueType *type)
 {
-  if (!is_py_obj_bytes_or_string_type(ret))
+  if (state->config_version_3_x_compat && !is_py_obj_bytes_or_string_type(ret))
     {
-      msg_error("$(python): The return value is not str or unicode",
+      msg_error("$(python): The current syslog-ng version does not support returning non-string values. "
+                "Later syslog-ng versions (at least 4.0) support it. "
+                "Please cast to string or consider upgrading the config version.",
```

Comment:
I would spell this a bit differently. 

```
    msg_error("$(python): The current config version does not support returning non-string values from Python functions. Please return str or bytes values from your Python function, use an explicit syslog-ng level cast to string() or set the config version to post 4.0"
                       cfg_format_config_version_tag(),
                       cfg_format_version_tag()
```

Please don't use a terminating dot at the end of a log message, it looks a bit funny when syslog-ng uses a semicolon between the message and the tags.

Reply:
Thanks, fixed.

I only used `cfg_format_config_version_tag()`, I do not know what the second tag `cfg_format_version_tag()` should hold in your example.

Reply:
I think I meant to include the value of FEATURE_TYPING_MIN_VERSION as the "4.0" version number would be different in PE.

e.g. cfg_format_version_tag("typing-min-version", FEATURE_TYPING_MIN_VERSION)

was my idea. It would be better to include that in the description though (in place of "post 4.0") but we lack the appropriate macro for that. I think the cfg_format_version_tag() based solution suffices